state/runtime/evm/state.go

```diff
@@ -210,30 +213,84 @@ func (c *state) resetReturnData() {
 
 // Run executes the virtual machine
 func (c *state) Run() ([]byte, error) {
-	var vmerr error
+	var (
+		vmerr  error
+		logged bool
+
+		op      OpCode
+		ipCopy  int
+		gasCopy uint64
+
+		ok bool
+	)
+
+	tracer := c.host.GetTracer()
+
+	defer func() {
+		if vmerr == nil || tracer == nil {
+			return
+		}
+
+		if !logged {
+			tracer.ExecuteState(
+				c.msg.Address,
+				ipCopy,
+				op.String(),
+				gasCopy,
+				// XXX: c.currentConsumedGas is not the total of gas assumption in a instruction
+				// when the execution terminated in the middle
+				c.currentConsumedGas,
+				c.returnData,
+				c.msg.Depth,
+				c.err,
+				c.host,
+			)
+		} else {
+			tracer.ExecuteFault(
+				ipCopy,
+				op.String(),
+				gasCopy,
+				// XXX: c.currentConsumedGas is not the total of gas assumption in a instruction
+				// when the execution terminated in the middle
+				c.currentConsumedGas,
+				c.msg.Depth,
+				c.err,
+			)
+		}
+	}()
 
-	codeSize := len(c.code)
 	for !c.stop {
-		if c.ip >= codeSize {
+		op, ok = c.CurrentOpCode()
+
+		if tracer != nil {
+			logged, ipCopy, gasCopy = false, c.ip, c.gas
+
+			// copy data before execution
+			tracer.CaptureMemory(c.memory)
```

Comment:
Is there any reason for having 3 calls instead of a single one?

Reply:
Actually no special reason and will merge these 3 methods

Reply:
Merged d730d2c56ed1aed10fe730b60d255a87293f458c